File: app/src/main/java/com/starsep/sokoban/gamelogic/HighScore.java
package com.starsep.sokoban.gamelogic;

import android.support.annotation.NonNull;

public class HighScore {
	public final int hash;
	public final int levelNumber;
	public int time;
	public int moves;
	public int pushes;
	public HighScore(int hash, int level, int time, int moves, int pushes) {
		this.hash = hash;
		levelNumber = level;
		this.time = time;
		this.moves = moves;
		this.pushes = pushes;
	}

	public void improve(@NonNull HighScore another) throws DifferentLevelsException {
		if (hash != another.hash) {
			throw new DifferentLevelsException();
		}
		time = Math.min(time, another.time);
		moves = Math.min(moves, another.moves);
		pushes = Math.min(pushes, another.pushes);
	}

	@Override
	@NonNull
	public String toString() {
		return "HighScore(" +
				"level=" + levelNumber + ", " +
				"hash=" + hash + ", " +
				"time=" + time + ", " +
				"moves=" + moves + ", " +
				"pushes=" + pushes + ")";
	}

	public class DifferentLevelsException extends Exception {}
}

File: app/src/main/java/com/starsep/sokoban/gamelogic/HighScore.kt
package com.starsep.sokoban.gamelogic

import android.arch.persistence.room.ColumnInfo
import android.arch.persistence.room.Entity

@Entity
data class HighScore(@ColumnInfo(name = "time") var time: Int = 0,
                     @ColumnInfo(name = "moves") var moves: Int = 0,
                     @ColumnInfo(name = "pushes") var pushes: Int = 0) {
    fun improve(another: HighScore) {
        time = Math.min(time, another.time)
        moves = Math.min(moves, another.moves)
        pushes = Math.min(pushes, another.pushes)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is HighScore) {
            return false
        }
        return time == other.time &&
                moves == other.moves &&
                pushes == other.pushes
    }

    override fun toString(): String {
        return "HighScore(" +
                "time=" + time + ", " +
                "movesLive=" + moves + ", " +
                "pushes=" + pushes + ")"
    }

    override fun hashCode(): Int {
        return toString().hashCode()
    }
}
